rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -150,3 +157,74 @@ impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
         internal_server_error()
     }
 }
+
+#[derive(Clone)]
+pub struct ServerRequestIdResponseProvider<S> {
+    inner: S,
+    header_key: HeaderName,
+}
+
+/// A layer that inserts the generated [`ServerRequestId`] to the response headers
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct ServerRequestIdResponseProviderLayer {
+    header_key: Box<str>,
+}
+
+impl ServerRequestIdResponseProviderLayer {
+    /// Add the request ID to the response header `header_key`
+    pub fn new(header_key: Box<str>) -> Self {
+        Self { header_key }
+    }
+}
+
+impl<S> Layer<S> for ServerRequestIdResponseProviderLayer {
+    type Service = ServerRequestIdResponseProvider<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        let header_key = HeaderName::from_bytes(self.header_key.as_bytes()).unwrap();
+        ServerRequestIdResponseProvider { inner, header_key }
+    }
+}
+
+impl<Body, S> Service<http::Request<Body>> for ServerRequestIdResponseProvider<S>
+where
+    S: Service<http::Request<Body>, Response = Response<crate::body::BoxBody>>,
+    S::Future: std::marker::Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = std::pin::Pin<Box<dyn Send + std::future::Future<Output = Result<Self::Response, Self::Error>>>>;
```

Comment:
NIT: Usually auto-traits come _after_ other traits - `dyn Future<...> + Send`.

Reply:
Also this is a heap allocation for something that can be done pretty easily via https://docs.rs/tower/latest/tower/trait.ServiceExt.html#method.map_response and then opaque type it.

Reply:
Wrong advice again, you can't easily opaque type because `request_id` is captured.